RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.cc

```diff
@@ -0,0 +1,499 @@
+      }
+
+    }
+  }
+
+  // hand over the final seeds to the output
+  // would be more elegant if we could do the above step with
+  // erasing the merged ones, rather than the
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(usedCluster[NNN]) continue; //skip seeds that have been marked as used up in merging
+    rechits_chains.push_back(seeds[NNN]);
+  }
+
+  //***************************************************************
+
+  return rechits_chains;
+}
+
+bool GEMSegAlgo::isGoodToMerge(const EnsembleHitContainer& newChain, const EnsembleHitContainer& oldChain) {
+
+  std::vector<float> phi_new, eta_new, phi_old, eta_old;
+  std::vector<int> layer_new, layer_old;
+  // std::vector< std::vector<int> > layer_new, layer_old;
+  std::vector<int> bx_new, bx_old;
+
+
+  // Do not allow to Merge if number of hits of Merged vector > 6
+  /*
+  if(newChain.size() + oldChain.size() > 6) {
+    edm::LogVerbatim("GEMSegAlgo") << "Merging those two vectors will lead to vector with more than 6 rechits :: n_vec1 = "<<newChain.size()<<" n_vec2 = "<<oldChain.size()<<" => Merging not allowed";
+    return false;
+  }
+  */
+
+  // edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::isGoodToMerge] :: newChain ::";
+  for(size_t iRH_new = 0;iRH_new<newChain.size();++iRH_new){
+    GEMDetId rhID                  = newChain[iRH_new]->gemId();
+    const GEMEtaPartition * rhEP   = (theEnsemble.second.find(rhID.rawId()))->second;
+    if(!rhEP) throw cms::Exception("GEMEtaPartition not found") << "Corresponding GEMEtaPartition to GEMDetId: "<<rhID<<" not found in the GEMEnsemble";
+    LocalPoint rhLP_inEtaPartFrame = newChain[iRH_new]->localPosition();
+    GlobalPoint rhGP_inCMSFrame    = rhEP->toGlobal(rhLP_inEtaPartFrame);
+    layer_new.push_back((rhID.station()-1)*2+rhID.layer());
+    // std::vector<int> layer_tmp; layer_tmp.push_back((rhID.station()-1)*2+rhID.layer()); layer_new.push_back(layer_tmp);
+    phi_new.push_back(rhGP_inCMSFrame.phi());
+    eta_new.push_back(rhGP_inCMSFrame.eta());
+    bx_new.push_back(newChain[iRH_new]->BunchX());
+    // Log Message
+    #ifdef EDM_ML_DEBUG // have lines below only compiled when in debug mode
+    const GEMSuperChamber * rhCH   = theEnsemble.first;
+    LocalPoint rhLP_inChamberFrame = rhCH->toLocal(rhGP_inCMSFrame);
+    edm::LogVerbatim("GEMSegAlgo") << "[GoodToMerge::New Chain][RecHit :: Loc x = "<<std::showpos<<std::setw(9)<<rhLP_inChamberFrame.x()<<" Loc y = "<<std::showpos<<std::setw(9)<<rhLP_inChamberFrame.y()
+    			             << " -- Glob eta = "<<std::showpos<<std::setw(9)<<rhGP_inCMSFrame.eta()<<" Glob phi = "<<std::showpos<<std::setw(9)<<rhGP_inCMSFrame.phi()
+                                     << " -- BX = "<<newChain[iRH_new]->BunchX()
+   	                 	     << " -- "<<rhID.rawId()<<" = "<<rhID<<" ]";
+    #endif
+  }  
+
+  // edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::isGoodToMerge] :: oldChain ::";
+  for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){
```

Comment:
if this extra debug  ``` edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::is ...``` is not needed, please remove it from the commits

... same for other places